testing/internal/e2e/tests/base/target_pagination_test.go

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package base_test
+
+import (
+	"context"
+	"encoding/json"
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/boundary/api/scopes"
+	"github.com/hashicorp/boundary/api/targets"
+	"github.com/hashicorp/boundary/globals"
+	"github.com/hashicorp/boundary/internal/target"
+	"github.com/hashicorp/boundary/testing/internal/e2e"
+	"github.com/hashicorp/boundary/testing/internal/e2e/boundary"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestCliPaginateTargets asserts that the CLI automatically paginates to retrieve
+// all targets in a single invocation.
+func TestCliPaginateTargets(t *testing.T) {
+	e2e.MaybeSkipTest(t)
+	c, err := loadTestConfig()
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	boundary.AuthenticateAdminCli(t, ctx)
+	newOrgId := boundary.CreateNewOrgCli(t, ctx)
+	t.Cleanup(func() {
+		ctx := context.Background()
+		boundary.AuthenticateAdminCli(t, ctx)
+		output := e2e.RunCommand(ctx, "boundary", e2e.WithArgs("scopes", "delete", "-id", newOrgId))
+		require.NoError(t, output.Err, string(output.Stderr))
+	})
+	newProjectId := boundary.CreateNewProjectCli(t, ctx, newOrgId)
+
+	// Create enough targets to overflow a single page.
+	// Use the API to make creation faster.
+	client, err := boundary.NewApiClient()
+	require.NoError(t, err)
+	tClient := targets.NewClient(client)
+	targetPort, err := strconv.ParseInt(c.TargetPort, 10, 32)
+	require.NoError(t, err)
+	for i := 0; i < globals.DefaultMaxPageSize+1; i++ {
+		_, err := tClient.Create(ctx, "tcp", newProjectId,
+			targets.WithName("test-target-"+strconv.Itoa(i)),
+			targets.WithTcpTargetDefaultPort(uint32(targetPort)),
+			targets.WithAddress(c.TargetAddress),
+		)
+		require.NoError(t, err)
+	}
+
+	// List targets recursively
+	output := e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"targets", "list",
+			"-scope-id", newProjectId,
+			"-format=json",
+		),
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+
+	var initialTargets targets.TargetListResult
+	err = json.Unmarshal(output.Stdout, &initialTargets)
+	require.NoError(t, err)
+
+	require.Len(t, initialTargets.Items, globals.DefaultMaxPageSize+1)
```

Comment:
I might be missing something conceptually here...

I thought that when using the list endpoint, we're only going to get `defaultMaxPageSize` items back from the endpoint and that we'd have to invoke another `list` to get the next page of items?

Reply:
This is the case over the API directly, yes, but the SDK will automatically paginate to the end, which means both the "Api" and "Cli" tests will get all results after the first call.